Add RFP-013 reflexive stablecoin and RFP-014 liquidation engine (#25)
logos-co/rfp: RFPs/RFP-013-reflexive-stablecoin-protocol.md

---
id: RFP-013
title: Reflexive Stablecoin Protocol
tier: XL
funding: $XXXXX
status: draft
dependencies: RFP-001 (Admin Authority), RFP-002 (Freeze Authority)
category: Applications & Integrations
---

# RFP-013 — Reflexive Stablecoin Protocol

## Overview

> Build a non-pegged, reflexive stablecoin protocol for the Logos Execution Zone (LEZ). The stablecoin is backed by a reference token collateral and uses an autonomous feedback controller to adjust a floating redemption price, incentivizing market participants to stabilize the token without requiring active governance intervention.

Reviewer: Should we specify that the token collateral must be LOGOS token?

Author: I intentionally left this open as "reference token", it doesn't make a difference imo and makes legal folks happy.

Reviewer: It makes sense.

The protocol follows the RAI design with these key components: collateralized debt positions (SAFEs) that track collateral, normalized debt, and accumulated rates; a redemption price that drifts based on a computed redemption rate; stability fees that accrue continuously via rate accumulation; and a feedback loop that stabilizes the token mathematically rather than through governance intervention.

## Why This Matters

The Logos ecosystem needs a credibly neutral settlement asset that reduces trust assumptions. Traditional stablecoins are either centralized (USDC), governance-heavy (DAI), or under-collateralized and fragile (failed algorithmic coins). This protocol provides a fourth option: reflexive, non-pegged, and over-collateralized.

> Stablecoins enable lending, trading, and payments without the volatility of native assets. For the private economy, this offers a stable unit of account without requiring users to trust an issuer or governance body. Other protocols can build on this stable asset as a risk-off collateral asset, establishing a foundation for DeFi activity on LEZ.

Reviewer: RAI is not as stable as USDC and DAI, and not as volatile as ETH, LOGOS. Is it worth distinguish it here?

Author: The goal is to describe how to build the system, not debate its merits against USDC or DAI. The comparison in 'Why This Matters' focuses on architectural differences, centralised vs autonomous governance, not downstream outcomes like volatility.

3. Debt accrues interest continuously via a stability fee mechanism. The fee rate is governance-adjustable but applies uniformly to all positions.

4. Positions cannot be modified in ways that would push their collateralization ratio below the liquidation threshold. Read-only queries are always permitted.

5. The stablecoin must be a standard fungible token on LEZ, transferable and composable with other protocols. The protocol handles minting and burning via integration with the LEZ Token Program.

6. The protocol reads market prices from a configurable price feed. The interface requires price data with staleness detection; the specific implementation is flexible.

#### Usability

1. Build the program using the SPEL framework, which generates the IDL and client code from the program definition.

2. Provide a Logos mini-app GUI with local build instructions, downloadable assets, and loadable in Logos app (Basecamp) via git repo.

> 3. Provide a CLI that covers core functionality of the program. The CLI may have fewer features than the GUI mini-app but must support all essential operations.

Reviewer: Shouldn't this be the opposite: the CLI exposes all features, while the GUI mini-app may have fewer features?

Author: This is consistent across all the RFPs (006, 007, 008 use the same wording). The rationale is that the mini-app is the primary user-facing deliverable and it should cover the full UX. The CLI is a developer/power-user tool that needs the essentials but doesn't need every UI flow (e.g., previews, privacy disclosures, analytics views). That said, if the team prefers CLI-first with a lighter GUI, that's fine too, I think.  The point is that at least one interface must cover everything.

Reviewer 2: Would we expect someone/us to run a node to watch the program and take action to arbitrage? if so, we could request the CLI/Daemon to have the necessary features.

If we dont expect us to run such a daemon, then I would leave it to developers to build their own and run it.

Author: Good point. We should expect keeper bots to exist. The CLI should expose the primitives (query rates, trigger updates, check positions), and the SDK makes it composable. Building an actual keeper daemon using the SDK should be rather easy.

4. Users can view current collateralization ratio, minimum safe ratio, and projected outcomes of operations before executing them.

5. Users can view current redemption rate, redemption price, and projected debt growth from stability fees.

> 6. The SDK supports both public account operations and private account operations via the standard deshield-interact-reshield pattern.

Reviewer 2: The SDK is referenced here as a constraint, but the template requires it to be an explicit deliverable. Suggest expanding this item:

```suggestion
6. Provide an SDK that can be used to build Logos modules for interacting with the program. The SDK must support both public account operations and private account operations via the standard deshield-interact-reshield pattern.
```

7. Failed operations return clear, actionable error messages indicating why the operation was rejected.

#### Reliability

1. All financial calculations use sufficient precision to prevent overflow. Critical invariants include: total stablecoin minted equals sum of all position debts; collateralization ratios are enforced after every modifying operation.

2. The feedback controller must include safeguards against integral windup and rate explosion.

3. If the price feed is stale or unavailable, rate updates are paused. Existing positions remain operational but new debt generation may be restricted.

> 4. Parameter updates (stability fee, controller gains, safety ratios) are gated through admin authority. Emergency circuit breaker functionality is available via freeze authority.

Reviewer: Who has the power of pressing these emergency buttons?

Author: This delegates to RFP-002 (Freeze Authority POC), which is listed as a dependency. The freeze authority mechanism is defined there — the stablecoin protocol would integrate with that authority rather than defining its own.

> 4. Parameter updates (stability fee, controller gains, safety ratios) are gated through admin authority. Emergency circuit breaker functionality is available via freeze authority.

Reviewer 2: Two distinct statements about two different authorities — suggest splitting into separate testable items:

```suggestion
4. Parameter updates (stability fee, controller gains, safety ratios) are gated through admin authority.

5. Emergency circuit breaker functionality is available via freeze authority.
```

#### Performance

1. Position operations and rate updates must fit within LEZ transaction compute limits with headroom for network conditions.

2. The protocol supports thousands of concurrent positions without exceeding compute limits on any operation.

3. Rate updates complete within a small number of blocks. Position operations complete within one block.

#### Supportability

1. The program is deployed and tested on LEZ devnet/testnet.

> 2. End-to-end integration tests run against a LEZ sequencer (standalone mode) and are included in CI.

Reviewer 2: Suggest requiring real proof generation in CI:

```suggestion
2. End-to-end integration tests run against a LEZ sequencer (standalone mode) with `RISC0_DEV_MODE=0` and are included in CI.
```

> 3. CI must be green on the default branch.

Reviewer: Specify "main branch"?

Author: This refers to the implementation team's repo, not the RFP repo itself. Let's leave them the benefit of calling the default one, however they want - "default branch" covers main, master, or whatever convention they follow.

> 4. Every hard requirement in Functionality, Usability, Reliability, and Performance has at least one corresponding test.

Reviewer: Should we define a minimum coverage?

Author: The RFP already says "every hard requirement has at least one corresponding test" (S4). That's effectively 100% requirement coverage, which is stronger than a percentage-based code coverage metric. Adding a code coverage number (e.g., 80%) on top of that is fine but somewhat redundant.

Reviewer: It depends on how we define the requirements. They usually only establish the happy path, and leave behind edge cases, safe guards, etc.. We can easily have 100% of the requirements tested with less than 50% of code coverage.

Author: Fair point, requirement coverage ≠ code coverage. But the requirements here aren't purely happy-path: they include error conditions (stale oracle → pause, insufficient collateral → reject, overflow protection, controller safeguards). We use the same approach across the other RFPs, so let's see how it plays out with the first submissions and adjust if needed.

#### + Privacy

1. The mini-app and SDK must support both direct public account interaction and the deshield-interact-reshield pattern for private account interaction. When a user chooses the private account path, the SDK must enforce the complete pattern — the reshield step must not be skippable.

2. When using the private account path, the SDK must validate that the target account for reshielding is a private (shielded) account before submitting the transaction, and reject the operation with an explicit error if it is not.

3. The ephemeral public account created during the deshield step must never be reused across operations. Each protocol interaction from a private account must use a freshly generated account with no prior on-chain history.

### Soft Requirements

If possible.

#### Reliability

1. Multi-oracle redundancy for price feeds, with fallback when primary is stale.

2. Formal verification of critical invariants: a position with collateralization ratio above the minimum cannot be liquidated; total stablecoin supply equals total debt across all positions.

#### Usability

> 1. Health factor preview showing before/after state for operations.

Reviewer: What's "health factor" here?

Author: Same as RFP-008: ratio of weighted collateral value to outstanding debt. HF ≥ 1 means solvent, HF < 1 means eligible for liquidation, rather standard terminology in defi.

> 2. Historical data display: redemption price, market price, and redemption rate over time.

Reviewer: Isn't this part of UI/UX?

Author: It is UI/UX, that's why it's under Soft Requirements → Usability. The core mini-app views (position state, collateral ratios, redemption rate) are hard requirements. Historical charting is nice-to-have but not essential for the protocol to function.

## Platform Dependencies

### Hard blockers

These must be available on LEZ before this RFP can open.

#### Price feed

The protocol requires external price feeds for market price and redemption rate computation. Every hard requirement related to rate adjustment (F2) and reliability (R3) depends on this.

#### On-chain clock / timestamp

Rate computation and stability fee accrual require knowing how much time has elapsed between interactions. Without a reliable on-chain timestamp, these cannot be computed.

### Soft blockers

Desirable but the RFP can open without them.

#### Event emission

Off-chain services need to monitor positions and rate updates. Without structured events, services must poll all accounts, which is expensive.

## Recommended Team Profile

Team experienced with:

- DeFi protocol design (CDP systems, RAI/Reflexer patterns)
- Solana or SVM program development
- Fixed-point arithmetic and numerical methods
- Interest rate modelling and control theory
- Oracle integration and price feed security
- Writing and running on-chain tests
- Front-end development for DeFi applications

## Timeline Expectations

> Estimated duration: **14–18 weeks**

Reviewer: Isn't this too much time? This is achievable with 14-18 days with Claude Code Max 20x..

Author: The timeline was benchmarked based on the lending protocol RFP scope. @fryorcraken can confirm the rationale here.

Reviewer 2: I am not able to justify the estimated durations I made. I think fine to make it shorter, ultimately, the dev studio should justify  longer time frames in their proposals.
